src/core/utils/jam_detector.cpp

```diff
@@ -0,0 +1,238 @@
+/*
+ *  Copyright (c) 2016, The OpenThread Authors.
+ *  All rights reserved.
+ *
+ *  Redistribution and use in source and binary forms, with or without
+ *  modification, are permitted provided that the following conditions are met:
+ *  1. Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *  2. Redistributions in binary form must reproduce the above copyright
+ *     notice, this list of conditions and the following disclaimer in the
+ *     documentation and/or other materials provided with the distribution.
+ *  3. Neither the name of the copyright holder nor the
+ *     names of its contributors may be used to endorse or promote products
+ *     derived from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ *  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ *  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ *  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ *  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ *  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ *  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * @file
+ *   This file implements the jam detector feature.
+ */
+
+#ifdef OPENTHREAD_CONFIG_FILE
+#include OPENTHREAD_CONFIG_FILE
+#else
+#include <openthread-config.h>
+#endif
+
+#include <openthread.h>
+#include <openthread-types.h>
+#include <thread/thread_netif.hpp>
+#include <common/code_utils.hpp>
+#include <platform/random.h>
+#include <utils/jam_detector.hpp>
+
+namespace Thread {
+namespace Utils {
+
+JamDetector::JamDetector(ThreadNetif &aNetif) :
+    mNetif(aNetif),
+    mTimer(aNetif.GetIp6().mTimerScheduler, &JamDetector::HandleTimer, this)
+{
+    mWindow = kMaxWindow;
+    mBusyPeriod = kMaxWindow;
+    mRssiThreshold = kDefaultRssiThreshold;
+
+    mEnabled = false;
+
+    mHandler = NULL;
+    mContext = NULL;
+
+    mHistoryBitmap = 0;
+}
+
+ThreadError JamDetector::Start(Handler aHandler, void *aContext)
+{
+    ThreadError error = kThreadError_None;
+
+    VerifyOrExit(!mEnabled, error = kThreadError_Already);
+    VerifyOrExit(aHandler != NULL, error = kThreadError_InvalidArgs);
+
+    mHandler = aHandler;
+    mContext = aContext;
+
+    mEnabled = true;
+
+    mCurSecondStartTime = Timer::GetNow();
+    mAlwaysAboveThreshold = true;
+    mHistoryBitmap = 0;
+    mJamState = false;
+    mSampleInterval = kMaxSampleInterval;
+
+    mTimer.Start(kMinSampleInterval);
```

Comment:
This won't work effectively if OpenThread is running on the host OS (on Windows). You can't guarantee such timer accuracy. This seems like an interesting feature that would be great to support on Windows, but would require an ability to offload some of this logic to the radio layer; similar to how we do it for the energy scan.